R/serializer.R

```diff
@@ -558,68 +573,142 @@ serializer_device <- function(type, dev_on, dev_off = grDevices::dev.off) {
       } else {
         success(result)
       }
-    }
+    },
+    serializer_params = serializer_params
   )
 }
 
+serializer_param_width <- function() {
+  list(plot_width = list(desc="Width of plot image in units", type="number", required=FALSE, isArray=FALSE))
+}
+serializer_param_height <- function() {
+  list(plot_height = list(desc="Height of plot image in units", type="number", required=FALSE, isArray=FALSE))
+}
+serializer_param_res <- function() {
+  list(plot_res = list(desc="Resolution of plot image", type="number", required=FALSE, isArray=FALSE))
+}
+serializer_param_units <- function() {
+  list(plot_units = list(desc="Units of plot image", type="string", required=FALSE, isArray=FALSE))
+}
+serializer_param_pointsize <- function() {
+  list(plot_pointsize = list(desc="Point size of plot image - TODO - better desc", type="number", required=FALSE, isArray=FALSE))
+}
+serializer_param_bg <- function() {
+  list(plot_bg = list(desc="Background colour of plot image", type="string", required=FALSE, isArray=FALSE))
+}
+
+serializer_param_list <- function(all_except = character()) {
+  params <- c(
+    serializer_param_width(),
+    serializer_param_height(),
+    serializer_param_res(),
+    serializer_param_units(),
+    serializer_param_pointsize(),
+    serializer_param_bg()
+  )
+
+  # oh I wish I had purrr!
+  ignored <- Map(function(x) {
+    params[[x]] <<- NULL
+  }, all_except)
+
+  params
+}
+
+
+serialize_dimensions_args_preparer <- function(req, ...) {
+
+  # if legacy plumber.staticSerializers is specified then ignore all req
+  # based parameters
+  if (getOption("staticSerializers", default="FALSE") == "TRUE") {
+    return(list(...))
+  }
+
+  as_numeric_nullable <- function(preferred, alternative) {
+    option <- preferred %||% alternative
+    if (is.null(option)) {
+      NULL
+    } else {
+      as.numeric(option)
+    }
+  }
+
+  doc_args <- list(...)
+  doc_args$width <- as_numeric_nullable(req$args$plot_width, doc_args$width)
+  doc_args$height <- as_numeric_nullable(req$args$plot_height, doc_args$height)
+  doc_args$units <- req$args$plot_units %||% doc_args$units
+  doc_args$res <- as_numeric_nullable(req$args$plot_res, doc_args$res)
+  doc_args$pointsize <- as_numeric_nullable(req$args$plot_pointsize, doc_args$pointsize)
+  doc_args$bg <- req$args$plot_bg %||% doc_args$bg
+
+  return(doc_args)
+}
+
+serializer_image_dev_on_func <- function(grFunc, ...) {
+  function(filename, req) {
+    dimension_args <- serialize_dimensions_args_preparer(req, ...)
+    rlang::exec(grFunc, filename, !!!dimension_args)
+  }
+}
+
+
 #' @describeIn serializers JPEG image serializer. See also: [grDevices::jpeg()]
 #' @export
 serializer_jpeg <- function(..., type = "image/jpeg") {
   serializer_device(
     type = type,
-    dev_on = function(filename) {
-      grDevices::jpeg(filename, ...)
-    }
+    dev_on = serializer_image_dev_on_func(grDevices::jpeg, ...),
+    serializer_params = serializer_param_list()
   )
 }
```

Comment:
We should allow users to determine if the dynamic params are used as the serializer level.

Maybe something like:

```r
get_serializer_params <- function(dynamic_params) {
  has_dynamic_params <- 
    rlang::maybe_missing(dynamic_params, {
      # Legacy support for static serializer parameters
      if (isTRUE(getOption("plumber.staticSerializers", default=FALSE)) return(FALSE)
      # Support dynamic params
      TRUE
    })
  if (!has_dynamic_params) return(list())
  serializer_param_list()
}
serializer_jpeg <- function(..., type = "image/jpeg", dynamic_params = missing_arg()) {
  serializer_device(
    type = type,
    dev_on = serializer_image_dev_on_func(grDevices::jpeg, ...),
    serializer_params = get_serializer_params(dynamic_params)
  )
}
```

Will need to document the `dynamic_params` parameter.